src/orderbook.ts

```diff
@@ -99,18 +139,93 @@ export class Orderbook {
       this.addAsk(ask);
     });
   }
+
+  /**
+   * Computes the transitive closure of this orderbook (e.g. ETH/DAI) with another one (e.g. DAI/USDC).
+   * Throws if the orderbooks cannot be combined (baseToken is not equal to quoteToken)
+   * @param orderbook The orderbook for which the transitive closure will be computed
+   * @returns A new instance of an orderbook representing the resulting closure.
+   */
+  transitiveClosure(orderbook: Orderbook) {
+    if (orderbook.baseToken != this.quoteToken) {
+      throw new Error(
+        `Cannot compute transitive closure of ${this.pair()} orderbook and ${orderbook.pair()} orderbook`
+      );
+    }
+
+    const ask_closure = this.transitiveAskClosure(orderbook);
+
+    // Since bids are the asks of the inverted orderbook, computing transitive closure of bids is equivalent to
+    // 1) inverting both orderbooks
+    // 2) computing the transitive ask closure on the inverses
+    // 3) re-inverting the result
+    const bid_closure = orderbook
+      .inverted()
+      .transitiveAskClosure(this.inverted())
+      .inverted();
+
+    ask_closure.add(bid_closure);
+    return ask_closure;
+  }
+
+  private transitiveAskClosure(orderbook: Orderbook) {
+    const result = new Orderbook(this.baseToken, orderbook.quoteToken);
+
+    // Create a copy here so original orders stay untouched
+    const left_asks = Array.from(this.asks.values());
+    const right_asks = Array.from(orderbook.asks.values());
+
+    left_asks.sort(sortOffersAscending);
+    right_asks.sort(sortOffersAscending);
+
+    const left_iterator = left_asks.values();
+    const right_iterator = right_asks.values();
+
+    let right_next = right_iterator.next();
+    let left_next = left_iterator.next();
+    while (!(left_next.done || right_next.done)) {
+      const right_offer = right_next.value;
+      const left_offer = left_next.value;
+      const price = left_offer.price.mul(right_offer.price);
+      let volume;
+      const right_offer_volume_in_left_offer_base_token = right_offer.volume.div(
+        left_offer.price
+      );
+      if (
+        left_offer.volume.toNumber() >
+        right_offer_volume_in_left_offer_base_token.toNumber()
+      ) {
+        volume = right_offer_volume_in_left_offer_base_token;
+        left_offer.volume = left_offer.volume.sub(volume);
+        right_next = right_iterator.next();
+      } else {
+        volume = left_offer.volume;
+        right_offer.volume = right_offer.volume.sub(
+          volume.mul(left_offer.price)
+        );
+        left_next = left_iterator.next();
+        // In case the orders matched perfectly we will move right as well
+        if (right_offer.volume.toNumber() == 0) {
+          right_next = right_iterator.next();
+        }
+      }
+      result.addAsk(new Offer(price, volume));
```

Comment:
Do fees need to be considered here somehow?

Reply:
Yes that is true. I left it our here initially as wrapping my head around the logic in the absence of a fee seemed simpler and already hard enough. I'm planning to implement this e2e without a fee for now but the later add a fee as a construction argument to the orderbook (or as a global param). We can then reduce the available amount when we multiply the asks on each round.